maintenance/measure_tile_performance/measure_tile_performance.py

```diff
@@ -15,24 +16,59 @@
 NAME = 'snowplow'
 HOST = 'redshift.analytics.gov.bc.ca'
 PORT = 5439
+user = os.environ['lookeruser_rs']
+passwd = os.environ['lookerpass_rs']
+connection_string = (
+    f"dbname='{NAME}' "
+    f"host='{HOST}' "
+    f"port='{PORT}' "
+    f"user='{user}' "
+    f"password={passwd}")
 
 # connects with redshift database using environment variables
 # set redshift cache to 'off' before running any looker API calls
 # run the API calls
 # set redshift cache back to 'on' after the queries are processed.
 # close the redshift connection
+
+def cache_off():
+ # set redshift cache to 'off'
+    conn = psycopg2.connect(dsn=connection_string)
+    with conn:
+         with conn.cursor() as curs:
+            try:
+                curs.execute("ALTER USER looker SET enable_result_cache_for_session TO off;")
+                print('redshift cache is off and connection is closed')
+            except Exception as err:
+                print(f'Exiting due to psycopg2 execution error: {err}')
+                print('redshift cache is on and connection is closed')
+                exit(1)
+
+def cache_on():
+ # set redshift cache back to 'on'
+    sleep_timer = 0
+    while True and sleep_timer <= 50: 
+        conn = psycopg2.connect(dsn=connection_string)
+        with conn:
+            with conn.cursor() as curs:
+                try:
+                    curs.execute("ALTER USER looker SET enable_result_cache_for_session TO on;")
+                    print('redshift cache is on and connection is closed')
+                except Exception as err:
+                    # exponential backoff loop to reconnect with database 5 times
+                    sleep_timer += 10
+                    if sleep_timer > 50:
+                        print(f'URGENT! After retrying redshift connection 5 times, program is exiting due to psycopg2 execution error: {err}')
+                        print('Requires manual run of: ALTER USER looker SET enable_result_cache_for_session TO on;')
+                        exit(1)
+                    else:    
+                        print(f"Retrying connection after {sleep_timer} seconds")
+                        time.sleep(sleep_timer)
```

Comment:
You should close the connection here before looping back and recreating a new connection object and assigning it to the `conn` variable. By doing so you've left the previous connection object alive and open, but you've just lost your `conn` reference to it since the `conn` variable gets assigned a new connection object on each loop.

Two reasons:
 1. The database will only allow a finite number of connections, so if one isn't required, is should be closed (sometimes the connection attempt succeeds, but the cursor execute attempt fails; which will still remain an open connection at the database end until you close the connection here).
 2. Python will have to wait until the connection times out before it will garbage collect each de-referenced connection object; this is unnecessary memory consumption.

From Psycopg2 docs: https://www.psycopg.org/docs/connection.html#connection 

> the connection is not closed by the context and it can be used for several contexts.

```python
conn = psycopg2.connect(DSN)

with conn:
    with conn.cursor() as curs:
        curs.execute(SQL1)

with conn:
    with conn.cursor() as curs:
        curs.execute(SQL2)

# leaving contexts doesn't close the connection
conn.close()
```

Reply:
SO the reason why i did not close the connection is beacuse it is automaically closed when the cursur is used in a with statement as you can read here.

https://www.psycopg.org/docs/cursor.html

_Changed in version 2.5: if the cursor is used in a with statement, the method is automatically called at the end of the with block._

IO tried closing it but it programs return the error that connection is already closed.

Reply:
https://www.psycopg.org/docs/cursor.html states that the cursor object closes at the end of its context (the `with` block) the connection is still open even at the end of its context, as stated in the documentation I linked to https://www.psycopg.org/docs/connection.html#connection.

Reply:
Closed the connection now.